Video gallery

dawoodkhan82 opened the pull request:
## Description
<img width="1254" alt="Screenshot 2024-08-07 at 12 48 40 PM" src="https://github.com/user-attachments/assets/15968589-0d1a-42fa-946f-ac44f4f157d3">
<img width="1286" alt="Screenshot 2024-08-07 at 12 48 45 PM" src="https://github.com/user-attachments/assets/4bb9be62-bc21-4913-bf5c-d5ea860b1423">

Closes: #2298

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

dawoodkhan82 commented:
Will add a param, to limit gallery uploads to only images or videos.

abidlabs commented on gradio/components/gallery.py:
```diff
@@ -31,24 +32,30 @@
 
+class GalleryVideo(GradioModel):
+    video: FileData
+    caption: Optional[str] = None
```
Do we support captions in videos in a Gallery? How would a user pass that in?

abidlabs commented on gradio/components/gallery.py:
```diff
@@ -31,24 +32,30 @@
+class GalleryVideo(GradioModel):
+    video: FileData
+    caption: Optional[str] = None
+    subtitles: Optional[FileData] = None
```
Do we support subtitles in videos in a Gallery? How would a user pass that in?

dawoodkhan82 replied:
Oh yeah will remove, there's no way to pass it in as of now. We can look into adding them later.

abidlabs commented:
Works quite well @dawoodkhan82! Just a few things I noticed:

(1) When I would click on a video, it would briefly show a circular loading icon in the center of the video. It feels unnecessary as nothing happens when the video finishes playing, can we just remove it?

https://github.com/user-attachments/assets/059fe291-1bf2-49ef-8435-a8d8b350bba1

(2) Captions don't show up for videos in the gallery mode:

<img width="795" alt="image" src="https://github.com/user-attachments/assets/3541a816-2077-4666-986b-aec89b4810f9">

They do show up when the video is selected though:

<img width="825" alt="image" src="https://github.com/user-attachments/assets/5c5d9800-913d-459f-81cf-fddb150a4210">

It'd be great if we can add the captions to the preview mode as well

> Will add a param, to limit gallery uploads to only images or videos.

Yeah, I think we can have a `file_types` parameter similar to `gr.File` that can be used to control images/videos or even more fine-grained control.

Happy to give it another pass and test events, etc. when ready

dawoodkhan82 commented:
@abidlabs 
- Removed subtitles (it didn't work anyways)
- Added `file_types` param. Defaults to allow both image & video. Maybe it should default to just image. thoughts?
- Fixed upload event bug
- Fixed caption on preview

> (1) When I would click on a video, it would briefly show a circular loading icon in the center of the video. It feels unnecessary as nothing happens when the video finishes playing, can we just remove it?

Not sure we can remove this, it's part of the `htmlvideoelement` controls. Also it only shows for a second or two while the video loads for me, then gets removed automatically.

abidlabs commented on gradio/components/gallery.py:
```diff
@@ -95,8 +102,9 @@ def __init__(
-            value: List of images to display in the gallery by default. If callable, the function will be called whenever the app loads to set the initial value of the component.
+            value: List of images or videos to display in the gallery by default. If callable, the function will be called whenever the app loads to set the initial value of the component.
             format: Format to save images before they are returned to the frontend, such as 'jpeg' or 'png'. This parameter only applies to images that are returned from the prediction function as numpy arrays or PIL Images. The format should be supported by the PIL library.
+            file_types: List of file extensions or types of files to be uploaded (e.g. ['image', 'video']). "image" allows only image files to be uploaded, "video" allows only video files to be uploaded, "video" allows only video files to be uploaded, ".mp4" allows only mp4 files to be uploaded, etc. If None, both image and video files types are allowed.
```
```suggestion
            file_types: List of file extensions or types of files to be uploaded (e.g. ['image', 'video']). "image" allows only image files to be uploaded, "video" allows only video files to be uploaded, ".mp4" allows only mp4 files to be uploaded, etc. If None, any image and video files types are allowed.
```

abidlabs commented on gradio/components/gallery.py:
```diff
@@ -170,27 +179,31 @@ def preprocess(
+            payload: a list of images or videos, or list of (media, caption) tuples
         Returns:
-            Passes the list of images as a list of (image, caption) tuples, or a list of (image, None) tuples if no captions are provided (which is usually the case). The image can be a `str` file path, a `numpy` array, or a `PIL.Image` object depending on `type`.
+            Passes the list of images as a list of (media, caption) tuples, or a list of (image, None) tuples if no captions are provided (which is usually the case). The image can be a `str` file path, a `numpy` array, or a `PIL.Image` object depending on `type`. The video can be a `str` file path.
```
```suggestion
            Passes the list of images or videos as a list of (media, caption) tuples, or a list of (media, None) tuples if no captions are provided (which is usually the case). Images can be a `str` file path, a `numpy` array, or a `PIL.Image` object depending on `type`.  Videos are always `str` file path.
```

abidlabs reviewed (approved):
LGTM, nice @dawoodkhan82! Reviewed the backend + functionality and everything looks good to me. Perhaps @aliabid94 @hannahblair or @pngwn can review the frontend code as well

freddyaboulton commented on js/gallery/shared/Gallery.svelte:
```diff
@@ -54,16 +56,30 @@
-
-	let prev_value: GalleryData | null = value;
+	let resolved_value: GalleryData[] | null = null;
+	let thumbnailsGenerated = false;
```
@dawoodkhan82 this implementation delivers but why do we need to manually generate thumbnails?

I think we can display videos inside the buttons like so

```svelte
<Video
  src={media.video.url}
  title={media.caption || null}
  is_stream={false}
  data-testid={"thumbnail " + (i + 1)}
  alt=""
  loading="lazy"
  loop={false}
/>
```

dawoodkhan82 replied:
Yeah originally had it this way, but thought loading videos would be too intensive. Looking at it again, i think you're right. Let me revert

freddyaboulton reviewed (commented):
@dawoodkhan82 thanks for making the change! Code looks good - one last night is that instead of having the `<Play>` button overlayed for videos, we can start playing the video on a mouseover? Like we do for video examples?

Tests are failing but it may be because of recent changes to 5.0-dev. Can take a look again once they pass.

freddyaboulton reviewed (approved):
LGTM once tests pass thanks @dawoodkhan82 !
